libhdfs3/src/rpc/RpcChannel.cpp

```diff
@@ -761,8 +761,8 @@ void RpcChannelImpl::buildConnectionContext(
     std::string euser = key.getAuth().getUser().getEffectiveUser();
     std::string ruser = key.getAuth().getUser().getRealUser();
 
-    if (!key.getAuth().getUser().hasEffectiveUser())
-        euser = principal;
+    if (!euser.empty())
```

Comment:
Not sure this change is correct. There is also still a call down below.

I played around with this a little today. I think this works if real and effective are both "user" but fails if effective is "user@DOMAIN" , even though that will get mapped to user also.

Reply:
I think behavior we want is:

1 Kerberos or simple, set effective to the effective user name if set, otherwise real. (for kerberos we get the real user name from the ticket).

2 For simple, if there is also a real user, set that.

This is a bit complicated, because I think in the UGI, real user name can be the effective user name and getUser is the original principal, so it won't match up to Java code exactly.